### PR TITLE
Format trailing where clauses in type aliases

### PR DESCRIPTION
**Pull request by compiler-errors**
Self-explanatory. cc rust-lang/rust#114901

This implementation is probably super busted. I have no idea what most of these options in rustfmt mean, lol.

### CODE REVIEW

**Comment by calebcartwright** [comment]
r? @ytmimi or @fee1-dead

I'd like to hold off on merging this till the next-next sync & release (assuming we are still on track to get let-chains in quickly) given the diff it'll introduce on stable.

Probably worth having common (and uncommon) comments in the test snippets too

**Comment by ytmimi** [comment]
> I'd like to hold off on merging this till the next-next sync & release (assuming we are still on track to get let-chains in quickly) given the diff it'll introduce on stable.

Sounds good. I think we're still on track to get that out soon. I'll likely have that PR ready later this week. Given that I'll probably hold off on reviewing this until after the next sync.

File: src/items.rs [hunk 1]
```diff
@@ -1657,6 +1657,9 @@ pub(crate) fn rewrite_type_alias<'a, 'b>(
         where_predicates_split,
     } = *ty_alias_kind;
     let ty_opt = ty.as_ref();
+    let rhs_hi = ty
+        .as_ref()
+        .map_or(where_clauses.0.1.hi(), |ty| ty.span.hi());
```
Comment (fee1-dead) [inline src/items.rs:1662]:
```suggestion
    let rhs_hi = ty_opt.map_or(where_clauses.0.1.hi(), |ty| ty.span.hi());
```

**Review by fee1-dead — APPROVED** [review]
implementation looks fine.

File: tests/target/type-alias-where-clauses.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+= E
+where
+    F: G,
+    H: I;
+
+type Foo
+    = E
+where
+    F: G,
+    H: I;
```
Comment (ytmimi) [inline tests/target/type-alias-where-clauses.rs:20]: @compiler-errors is there documentation in the style guide that you can link to for this?

I'm wondering why this wouldn't be formatted like:

```rust
type Foo = E
where
    F: G,
    H: I;
```

File: tests/target/type-alias-where-clauses.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+= E;
+
+type Foo
+where
+    A: B,
+    C: D,
+= E
+where
+    F: G,
+    H: I;
```
Comment (ytmimi) [inline tests/target/type-alias-where-clauses.rs:14]: This seems like an interesting case. Is it valid to have two where clauses?

Comment (compiler-errors) [reply tests/target/type-alias-where-clauses.rs:14]: No, but it's not denied by the parser, so I think rustfmt shouldn't explode when we have it.

File: tests/target/type-alias-where-clauses.rs (hunk 2 above)
Comment (compiler-errors) [reply tests/target/type-alias-where-clauses.rs:20]: https://github.com/rust-lang/rust/pull/114901

Comment (ytmimi) [reply tests/target/type-alias-where-clauses.rs:20]: Thank you. That's exactly what I was looking for

File: tests/target/type-alias-where-clauses.rs (hunk 3 above)
Comment (ytmimi) [reply tests/target/type-alias-where-clauses.rs:14]: I appreciate the clarification on this

Comment (calebcartwright) [reply tests/target/type-alias-where-clauses.rs:14]: > so I think rustfmt shouldn't explode when we have it.

I'd like to adopt this as a new motto for us 🤣

**Comment by calebcartwright** [comment]
@ytmimi  do you think we should pull this in as part of the next release? I imagine this would need to be prefaced with a blog post and some associated lead time due to the intro of formatting being applied for the first time?

**Review by ytmimi — APPROVED** [review]
I've been taking a look at this for a while now and the implementation makes sense to me. I was trying to think of ways to simplify / improve what you've got here, but I don't think I can come up with any. I think we're good to move forward with the PR

**Comment by ytmimi** [comment]
> @ytmimi do you think we should pull this in as part of the next release? I imagine this would need to be prefaced with a blog post and some associated lead time due to the intro of formatting being applied for the first time?

@calebcartwright Yeah, I think we could pull this in for the next release. When did you want to do the sync? I'm still up for taking on the `subtree push`. Maybe we can talk it over on Zulip and find a time that works.

**Comment by ytmimi** [comment]
I'm expecting to see some changes now that we actually format trailing where clauses, but let's see what the [Diff-Check](https://github.com/rust-lang/rustfmt/actions/runs/7661811341/job/20881893858) reveals.

**Comment by ytmimi** [comment]
Okay, This [Diff-Check](https://github.com/rust-lang/rustfmt/actions/runs/7661954853/job/20882324738) job shows the diffs I was expecting. @compiler-errors can when you have a chance, can you review the logs and double check that things are formatted here as expected based on the formatting rules you implemented.

For your convenience I'm including the diffs inline:
**note** the output is a diff of diffs, so it might be a little noisy, by the highlighting should help identify the changes


```diff
Diff in /tmp/rust-lang-rust-f0yK2vou/library/core/src/ops/async_function.rs:73:
     where
         <F as FnOnce<A>>::Output: Future,
     {
-        type CallFuture<'a> = <F as FnOnce<A>>::Output where Self: 'a;
+        type CallFuture<'a>
+            = <F as FnOnce<A>>::Output
+        where
+            Self: 'a;
 
         extern "rust-call" fn async_call(&self, args: A) -> Self::CallFuture<'_> {
             self.call(args)

Diff in /tmp/rust-lang-rust-f0yK2vou/library/core/src/ops/async_function.rs:85:
     where
         <F as FnOnce<A>>::Output: Future,
     {
-        type CallMutFuture<'a> = <F as FnOnce<A>>::Output where Self: 'a;
+        type CallMutFuture<'a>
+            = <F as FnOnce<A>>::Output
+        where
+            Self: 'a;
 
         extern "rust-call" fn async_call_mut(&mut self, args: A) -> Self::CallMutFuture<'_> {
             self.call_mut(args)

Diff in /tmp/rust-lang-rust-f0yK2vou/compiler/rustc_query_system/src/query/caches.rs:32:
 pub struct DefaultCacheSelector<K>(PhantomData<K>);
 
 impl<'tcx, K: Eq + Hash, V: 'tcx> CacheSelector<'tcx, V> for DefaultCacheSelector<K> {
-    type Cache = DefaultCache<K, V>
+    type Cache
+        = DefaultCache<K, V>
     where
         V: Copy;
 }

Diff in /tmp/rust-lang-rust-f0yK2vou/compiler/rustc_query_system/src/query/caches.rs:84:
 pub struct SingleCacheSelector;
 
 impl<'tcx, V: 'tcx> CacheSelector<'tcx, V> for SingleCacheSelector {
-    type Cache = SingleCache<V>
+    type Cache
+        = SingleCache<V>
     where
         V: Copy;
 }

Diff in /tmp/rust-lang-rust-f0yK2vou/compiler/rustc_query_system/src/query/caches.rs:126:
 pub struct VecCacheSelector<K>(PhantomData<K>);
 
 impl<'tcx, K: Idx, V: 'tcx> CacheSelector<'tcx, V> for VecCacheSelector<K> {
-    type Cache = VecCache<K, V>
+    type Cache
+        = VecCache<K, V>
     where
         V: Copy;
 }

Diff in /tmp/rust-lang-rust-f0yK2vou/compiler/rustc_query_system/src/query/caches.rs:177:
 pub struct DefIdCacheSelector;
 
 impl<'tcx, V: 'tcx> CacheSelector<'tcx, V> for DefIdCacheSelector {
-    type Cache = DefIdCache<V>
+    type Cache
+        = DefIdCache<V>
     where
         V: Copy;
 }

Diff in /tmp/rust-lang-rust-f0yK2vou/src/librustdoc/clean/types.rs:1059:
 }
 
 impl AttributesExt for [(Cow<'_, ast::Attribute>, Option<DefId>)] {
-    type AttributeIterator<'a> = impl Iterator<Item = ast::NestedMetaItem> + 'a
-        where Self: 'a;
-    type Attributes<'a> = impl Iterator<Item = &'a ast::Attribute> + 'a
-        where Self: 'a;
+    type AttributeIterator<'a>
+        = impl Iterator<Item = ast::NestedMetaItem> + 'a
+    where
+        Self: 'a;
+    type Attributes<'a>
+        = impl Iterator<Item = &'a ast::Attribute> + 'a
+    where
+        Self: 'a;
 
     fn lists<'a>(&'a self, name: Symbol) -> Self::AttributeIterator<'a> {
         AttributesExt::iter(self)
@ -7[54](https://github.com/rust-lang/rustfmt/actions/runs/7661954853/job/20882324738#step:4:55)9,0 +7[63](https://github.com/rust-lang/rustfmt/actions/runs/7661954853/job/20882324738#step:4:64)3,7 @@ Diff in /tmp/rust-lang-rust-f0yK2vou/src/tools/rust-analyzer/crates/parser/test_

Diff in /tmp/rust-lang-rust-f0yK2vou/src/tools/rust-analyzer/crates/parser/test_data/parser/inline/ok/0012_type_item_where_clause.rs:1:
-type Foo = () where Foo: Copy;
+type Foo
+    = ()
+where
+    Foo: Copy;
```

**Comment by compiler-errors** [comment]
@ytmimi: Yes, that formatting looks correct. When there is a where clause present, there's always a line break, then `=`, then a break, then the where clauses.

**Comment by ytmimi** [comment]
> @ytmimi do you think we should pull this in as part of the next release? I imagine this would need to be prefaced with a blog post and some associated lead time due to the intro of formatting being applied for the first time?

I think all the implementation work for this is done on the rustfmt side!

@calebcartwright I missed the fact that you suggested we get a blog post together for this. I'd be happy to draft something in preparation for the release. Is the idea to have the blog post go out before the release?

**Comment by ytmimi** [comment]
> @ytmimi: Yes, that formatting looks correct. When there is a where clause present, there's always a line break, then `=` and the type, then a break, then the where clauses.

Awesome, thank you for confirming! I'm already anticipating someone asking for a `single-line` formatting option 😅

**Comment by compiler-errors** [comment]
@calebcartwright @ytmimi: It's been almost a year since this PR was opened. Is there any chance we could finally get this merged? What's the plan on doing the next rustfmt->rust sync?

**Comment by calebcartwright** [comment]
@ytmimi thoughts on how to label this one? it isn't technically part of the 2024 edition and can/should go before, but it is definitely something that can't slip _past_ the 2024 edition

**Comment by ytmimi** [comment]
Agreed that this one should get in no later than the 2024 edition, and really as soon as we can. Just brainstorming here, but maybe we can label it something like `pre-2024-must-have`

File: src/items.rs [hunk 4]
```diff
@@ -1811,12 +1825,36 @@ fn rewrite_ty<R: Rewrite>(
             _ => format!("{result}="),
         };
 
-        // 1 = `;`
-        let shape = Shape::indented(indent, context.config).sub_width(1)?;
-        rewrite_assign_rhs(context, lhs, &*ty, &RhsAssignKind::Ty, shape).map(|s| s + ";")
+        // 1 = `;` unless there's a trailing where clause
+        let shape = if after_where_predicates.is_empty() {
+            Shape::indented(indent, context.config).sub_width(1)?
```
Comment (fmease) [inline src/items.rs:1830]:
```suggestion
        let shape = if after_where_predicates.is_empty() {
            Shape::indented(indent, context.config).sub_width(const { ";".len() })?
```

/half serious

Comment (compiler-errors) [reply src/items.rs:1830]: would rather do that for all the magic numbers in rustfmt all at once

Comment (calebcartwright) [reply src/items.rs:1830]: the width adjustment numbers aren't always based off a static string, and i'd rather stick with direct numbers and inline comments explaining where it's coming from, if for no reason other than succinctness

**Comment by calebcartwright** [comment]
What I think I'd like to do is get this merged such that it's timed to land with the addition of the `style_edition` option so that we can have a single blog post highlighting both

**Comment by calebcartwright** [comment]
@compiler-errors i hate to ask but would you have bandwidth to rebase? if not i understand (i know how busy you are) and i can grab your commits and pull them into a new branch but just wanted to check with you first

**Comment by compiler-errors** [comment]
@calebcartwright: Done. Sorry, was asleep by the time you sent that message otherwise I'd've done it sooner 😆

**Comment by compiler-errors** [comment]
Do you want me to rebase?

**Comment by compiler-errors** [comment]
Well, I went ahead and did it anyways.

**Comment by calebcartwright** [comment]
Thank you :pray:

**Pull request merged**